docs/design/provisioning.md

```diff
@@ -0,0 +1,328 @@
+// within. The Recycler controller running in master will use the specified plugin and optional arguments.
+// Recyclers can be plugins that delete cloud volumes through APIs or pods that scrub a volume for reuse.
+// Each Recycler plugin requires its own documentation for usage and plugin arguments.
+type RecyclerConfig struct {
+	// PersistentVolumeSelector is a selector which must be true for the recycler to match the persistent volume
+	PersistentVolumeSelector Selector
+	// RecyclerPluginName is the name of the plugin that can recycle the matched persistent volume.
+	RecyclerPluginName string
+	// Args is a map of key/value pairs passed to the driver (but not plugin) as arguments
+	Args *map[string]string
+	// SecurityContextName is the name of the security context to use when performing recycling operations
+	SecurityContextName string
+}
+
+// ProvisionerConfig is a label set for persistent volume claims, a plugin w/ optional arguments,
+// and a security context to work within.  The provisioning controller running in master will use the specified
+// plugin to provision volumes for claims that match the selector.
+// Each Provisioner plugin requires its own documentation for usage and plugin arguments.
+type ProvisionerConfig struct {
+	// PersistentVolumeClaimSelector is a selector which must be true for the provisioner to match the claim
+	PersistentVolumeClaimSelector Selector
+	// ProvisionerPluginName is the name of the plugin that will provision the claim matched by the selector
+	ProvisionerPluginName string
+	// Args is a map of key/value pairs passed to the driver (but not plugin) as arguments
+	Args *map[string]string
+	// SecurityContextName is the name of the security context to use when performing provisioning operations
+	SecurityContextName string
+}
+
+```
+
+### Provisioning Plugins
+
+New plugins:
+
+`PersistentVolumeProvisioner` is a plugin that can create real storage with a storage provider (such as EBS or GCE).
+
+`PersistentVolumeRecycler` is a plugin that reclaims discarded resources.
+
+> The Deleter plugin and interfaces to be removed. Deletion is one of many possible Recycling outcomes.
+> The Creater plugin and interfaces to be refactored to Provisioner
+
+
+Provisioning and Recycling both follow a plugin model where the plugin accepts  generic arguments.  A future enhancement can scan for drivers by name and consume the same arguments (following a similar pattern in  [Networking](../../pkg/kubelet/network/exec/exec.go)).
+
+The plugin model is similar to volume plugins, where each has a unique name and implements various interfaces.  Arguments as passed as string key/value pairs to a plugin in `Init(host, args)`, which will be called exactly once before any plugin operation is invoked.  Many provisioners can be configured using the same plugin with different arguments, such as "volumeType" for a Cloud API.
```

Comment:
Also, to your point about volume extensibility generally, there's FlexVolume (https://github.com/kubernetes/kubernetes/pull/13840) in the works to allow the same driver model to exec safely on a node to setup $FOO volume that's not compiled into Kube.